base/pkg/read.jl

```diff
@@ -2,15 +2,70 @@
 
 module Read
 
-import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir
+import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir, ..isdevmetadata
 using ..Types
 
+devmetapkgpath(s::AbstractString) = joinpath(uppercase(string(s[1])), s * ".versions")
+
 readstrip(path...) = strip(readstring(joinpath(path...)))
 
-url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
-sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
+function url(pkg::AbstractString, dev::Bool = false)
+    if dev
+        chomp(readline(joinpath(Dir.path("METADATA"), devmetapkgpath(pkg))))
+    else
+        readstrip(Dir.path("METADATA"), pkg, "url")
+    end
+end
+
+# Should only be called when cache is guaranteed to be updated
```

Comment:
I suspect a lot of these questions are going to be raised about the Pkg3 proposal as well. The question is how far are we going to move away from METADATA, and possibly packages themselves, being active git repositories on the filesystem. If you're not going to use the information in the working copy on the filesystem, then why have a copy there at all? It's misleading to have it there if changes made there don't actually matter. What's read-only and what is the workflow for making edits? When PkgDev doesn't work, as it often doesn't for a large number of people (yes we should fix that, but I don't know if we can realistically get it to 100% perfect in all git/auth configurations), what is the fallback option?